src/base/QXmppHttpUploadIq.cpp

```diff
@@ -0,0 +1,262 @@
+    d->fileName = request.attribute("filename");
+    d->size = request.attribute("size").toLongLong();
+    if (request.hasAttribute("content-type")) {
+        QMimeDatabase mimeDb;
+        QMimeType type = mimeDb.mimeTypeForName(request.attribute("content-type"));
+        if (!type.isDefault() && type.isValid())
+            d->contentType = type;
+    }
+}
+
+void QXmppHttpUploadRequestIq::toXmlElementFromChild(QXmlStreamWriter *writer) const
+{
+    writer->writeStartElement("request");
+    writer->writeAttribute("xmlns", ns_http_upload);
+    // filename and size are required
+    writer->writeAttribute("filename", d->fileName);
+    writer->writeAttribute("size", QString::number(d->size));
+    // content-type is optional
+    if (!d->contentType.isDefault() && d->contentType.isValid())
+        writer->writeAttribute("content-type", d->contentType.name());
+    writer->writeEndElement();
+}
+/// \endcond
+
+class QXmppHttpUploadSlotIqPrivate
+{
+public:
+    QUrl putUrl;
+    QUrl getUrl;
+    QMap<QString, QString> headerFields;
+};
+
+QXmppHttpUploadSlotIq::QXmppHttpUploadSlotIq()
+    : d(new QXmppHttpUploadSlotIqPrivate())
+{
+}
+
+QXmppHttpUploadSlotIq::~QXmppHttpUploadSlotIq()
+{
+    delete d;
+}
+
+/// Returns the URL for uploading via. HTTP PUT.
+
+QUrl QXmppHttpUploadSlotIq::putUrl() const
+{
+    return d->putUrl;
+}
+
+/// Sets the URL the client should use for uploading.
+
+void QXmppHttpUploadSlotIq::setPutUrl(const QUrl &putUrl)
+{
+    d->putUrl = putUrl;
+}
+
+/// Returns the URL to where the file will be served.
+
+QUrl QXmppHttpUploadSlotIq::getUrl() const
+{
+    return d->getUrl;
+}
+
+/// Sets the download URL.
+
+void QXmppHttpUploadSlotIq::setGetUrl(const QUrl &getUrl)
+{
+    d->getUrl = getUrl;
+}
+
+/// Returns a map of header fields (header name -> content) that need to be
+/// included in the PUT (upload) request. This won't contain any other fields
+/// than: "Authorization", "Cookie" or "Expires".
+
+QMap<QString, QString> QXmppHttpUploadSlotIq::headerFields() const
+{
+    return d->headerFields;
+}
+
+/// Sets the header fields the client needs to include in the PUT (upload)
+/// request. All fields other than "Authorization", "Cookie" or "Expires" will
+/// be ignored.
+
+void QXmppHttpUploadSlotIq::setHeaderFields(const QMap<QString, QString> &headerFields)
+{
+    d->headerFields.clear();
+    for (QString &name : headerFields.keys()) {
+        if (name == "Authorization" || name == "Cookie" || name == "Expires") {
+            QString value = headerFields[name];
+            d->headerFields[name] = value.replace("\n", "");
+        }
+    }
+}
+
+/// Returns whether the URLs (put and get) are both HTTPS URLs. It will return
+/// false, if the URLs are non-HTTPS URLs or not set. HTTP URLs are not
+/// XEP-compliant, but are used in some cases.
+
+bool QXmppHttpUploadSlotIq::hasHttpsUrls() const
+{
+    return d->getUrl.toString().startsWith("https://") &&
```

Comment:
Do we need this method? If we absolutely do, why not use url.scheme() ?

Reply:
A valid slot needs both URLs to be set. But you're right, the behaviour/https-checking can be done in the manager (and yes `.scheme()` were better of course :smile:).